readme.md
# eta-help

It is a help tool that collects feedbacks from users.

debian/changelog
eta-help (0.2.8) unstable; urgency=medium

  * Fix retry button

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 18 Oct 2018 08:21:45 +0000

eta-help (0.2.7) unstable; urgency=medium

  * Add screenshot animation

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 18 Oct 2018 07:55:47 +0000

eta-help (0.2.6) unstable; urgency=medium

  * Pardus CI integration

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 18 Oct 2018 07:25:27 +0000

eta-help (0.2.5) unstable; urgency=medium

  * Fix build dependency

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 18 Oct 2018 07:22:01 +0000

eta-help (0.2.4) unstable; urgency=medium

  * Change design

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 18 Oct 2018 07:19:18 +0000

eta-help (0.2.3) unstable; urgency=medium

  * Deleted make file

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 05 Oct 2017 07:05:12 +0000

eta-help (0.2.2) unstable; urgency=medium

  * Change user guide name

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Tue, 03 Oct 2017 12:10:07 +0000

eta-help (0.2.1) unstable; urgency=medium

  * Change binary target from etahelp to eta-help

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Tue, 03 Oct 2017 11:56:28 +0000

eta-help (0.2.0) unstable; urgency=medium

  * Initial release.

 -- Hikmet Baş <hikmet.bas@pardus.org.tr>  Thu, 14 Sep 2017 14:31:37 +0300


src/bridge.h
/*****************************************************************************
 *   Copyright (C) 2016 by Hikmet Bas                                        *
 *   <hikmet.bas@pardus.org.tr>                                              *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#ifndef BRIDGE_H
#define BRIDGE_H

#include <QObject>
#include <QString>

class QDBusInterface;


class RequestManager;
class Bridge : public QObject
{
    Q_OBJECT
    Q_PROPERTY(QString server READ server
               NOTIFY sent
               NOTIFY failed)
public:
    explicit Bridge(QObject *parent = 0);
    QString server() const;
    Q_INVOKABLE void sendMail(const QString &from, const QString &subject,
                              const QString &body, const QString &filePath);
    Q_INVOKABLE void sendMail(const QString &from, const QString &subject,
                              const QString &body);
    Q_INVOKABLE QString getScreenShot() const;
    Q_INVOKABLE void showKeyboard();
    Q_INVOKABLE void hideKeyboard();    
private:    
    QString srv;
    QDBusInterface *interface;    
    RequestManager *rm;
signals:
    void sent();
    void failed();
public slots:
    void sentSlot(const bool &);
};

#endif // BRIDGE_H


src/main.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Hikmet Bas                                        *
 *   <hikmet.bas@pardus.org.tr>                                              *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#include <QApplication>
#include <QQmlApplicationEngine>
#include <QtQml>
#include <QIcon>
#include "bridge.h"
#define ICONPATH "/usr/share/eta/eta-help/icon/help.svg"

int main(int argc, char *argv[])
{

    qmlRegisterType<Bridge>("eta.smtp",1,0,"EtaSmtp");
    QApplication::setWindowIcon( QIcon(ICONPATH) );
    QApplication app(argc, argv);
    QQmlApplicationEngine engine;
    engine.load(QUrl(QStringLiteral("qrc:/ui/main.qml")));
    return app.exec();
}


js/functions.js
/*****************************************************************************
 *   Copyright (C) 2016 by Hikmet Bas                                        *
 *   <hikmet.bas@pardus.org.tr>                                              *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

function checkMail(mail) {
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return re.test(mail)
}

function checkText(text) {
    var re= /^\s+$|^$/

    return re.test(text)
}


src/filemanager.h
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#ifndef FILEMANAGER_H
#define FILEMANAGER_H

#include <QObject>
#include <QJsonValue>
#include <QString>

class FileManager : public QObject
{
    Q_OBJECT
public:
    explicit FileManager(QObject *parent = 0);

signals:

public slots:
    QJsonValue convertFile(const QString& filePath) const;
    QString getFileName(const QString& filePath) const;

};

#endif // FILEMANAGER_H


src/filemanager.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#include "filemanager.h"
#include <QTextStream>
#include <QFile>
#include <QFileInfo>
#include <QByteArray>

FileManager::FileManager(QObject *parent) :
    QObject(parent)
{

}


QJsonValue FileManager::convertFile(const QString& filePath) const
{
    QByteArray bytes;
    QFile file(filePath);
    if(file.exists()) {
        if (!file.open(QIODevice::ReadOnly)) {
            qDebug("Couldn't open the file");
        } else {
            bytes = file.readAll();
        }

    } else {
        qDebug("Couldn't find the file");
    }
    return QJsonValue(QString::fromLatin1(bytes.toBase64()));
}

QString FileManager::getFileName(const QString &filePath) const
{
    QString fn = "unknown";
    QFile file(filePath);
    if(file.exists()) {
        fn = QFileInfo(file.fileName()).fileName();
    } else {
        qDebug("Couldn't find the file");
    }
    return fn;
}


src/requestmanager.h
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#ifndef REQUESTMANAGER_H
#define REQUESTMANAGER_H

#include <QObject>

class QNetworkReply;
class QNetworkAccessManager;
class QNetworkRequest;
class FileManager;

#define URL "http://etahelp.etap.org.tr/etahelp.php"

class RequestManager : public QObject
{
    Q_OBJECT
public:
    explicit RequestManager(QObject *parent = 0);

private:
    QNetworkAccessManager *nam;
    QNetworkRequest *request;
    FileManager *fm;

signals:
    void gotResult(bool b);

public slots:
    void requestFinished(QNetworkReply *r);
    void sendMail(const QString &from, const QString &subject,
                  const QString& details, const QString &filePath);

};

#endif // REQUESTMANAGER_H


src/requestmanager.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#include "requestmanager.h"
#include "filemanager.h"
#include <QNetworkRequest>
#include <QUrl>
#include <QJsonObject>
#include <QNetworkAccessManager>
#include <QJsonDocument>
#include <QNetworkReply>
#include <QJsonValue>
#include <QDebug>


RequestManager::RequestManager(QObject *parent) :
    QObject(parent)
{
    nam = new QNetworkAccessManager(this);
    fm = new FileManager(this);
    request = new QNetworkRequest(QUrl(URL));
    request->setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    connect(nam, SIGNAL(finished(QNetworkReply*)),
            this, SLOT(requestFinished(QNetworkReply*)));
}



void RequestManager::sendMail(const QString& from, const QString& subject,
                              const QString& details, const QString& filePath)
{
    QJsonObject data;
    data.insert("from", from);
    data.insert("subject", subject);
    data.insert("details", details);
    data.insert("filename", fm->getFileName(filePath));
    data.insert("attach",fm->convertFile(filePath));
    data.insert("status","");
    nam->post(*request, QJsonDocument(data).toJson());
}

void RequestManager::requestFinished(QNetworkReply *r)
{
    QByteArray response_data = r->readAll();
    QJsonDocument response_doc = QJsonDocument::fromJson(response_data);
    QJsonObject response = response_doc.object();
    bool b = response.value("status").toBool();
    emit gotResult(b);
}


src/bridge.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Hikmet Bas                                        *
 *   <hikmet.bas@pardus.org.tr>                                              *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/

#include "bridge.h"
#include "requestmanager.h"
#include <QPixmap>
#include <QDir>
#include <QScreen>
#include <QGuiApplication>
#include <QDBusConnection>
#include <QDBusInterface>

Bridge::Bridge(QObject *parent) :
    QObject(parent)
{
    srv = "etdestek";
    rm = new RequestManager(this);
    connect(rm,SIGNAL(gotResult(bool)),this,SLOT(sentSlot(bool)));
    QDBusConnection dbus = QDBusConnection::sessionBus();
    interface = new QDBusInterface("org.eta.virtualkeyboard",
                                   "/VirtualKeyboard",
                                   "org.eta.virtualkeyboard",
                                   dbus,
                                   this);
}

void Bridge::showKeyboard()
{
    interface->call("showFromBottom");
}

void Bridge::hideKeyboard()
{
    interface->call("hide");
}

QString Bridge::server() const
{
    return srv;
}
void Bridge::sendMail(const QString &from, const QString &subject,
                      const QString &body, const QString &filePath)
{    
    rm->sendMail(from,subject,body,filePath);
}

void Bridge::sendMail(const QString &from, const QString &subject,
                      const QString &body)
{
    const QString emptyPath = "";
    rm->sendMail(from,subject,body,emptyPath);
}

void Bridge::sentSlot(const bool &b)
{
    if (b) {
        emit sent();
    } else {
        emit failed();
    }
}

QString Bridge::getScreenShot() const
{
    QPixmap qPixMap;
    QScreen *screen = QGuiApplication::primaryScreen();
       if (screen)
           qPixMap = screen->grabWindow(0);    
    QString format = "jpg";
    QDir::setCurrent("/tmp");
    if(!QDir("Folder").exists())
      QDir().mkdir("eta-help");
    QDir::setCurrent("/tmp/eta-help");
    QString initialPath = QDir::currentPath() + tr("/screen.") + format;
    qPixMap.save(initialPath, format.toLatin1());

    return initialPath;
}


